File: backend/src/services/evidenceRevealMapping.ts
/**
 * Map model-produced revealedEvidence strings onto canonical hiddenEvidence titles
 * so chips/collection resolve to real Evidence rows (image + description).
 */

export interface EvidenceTitleMatch {
  id: string;
  title: string;
  description?: string;
}

export function stripRevealedEvidenceMetadata(line: string): string {
  const s = line.trim();
  const idx = s.search(/\s*\|\s*DISCOVERY_ZONE\b/i);
  return idx >= 0 ? s.slice(0, idx).trim() : s;
}

function levenshtein(a: string, b: string): number {
  if (a === b) return 0;
  const m = a.length;
  const n = b.length;
  if (m === 0) return n;
  if (n === 0) return m;
  const row = new Array<number>(n + 1);
  for (let j = 0; j <= n; j++) row[j] = j;
  for (let i = 1; i <= m; i++) {
    let prev = i - 1;
    row[0] = i;
    for (let j = 1; j <= n; j++) {
      const tmp = row[j];
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      row[j] = Math.min(row[j] + 1, row[j - 1] + 1, prev + cost);
      prev = tmp;
    }
  }
  return row[n];
}

function levSimilarity(a: string, b: string): number {
  if (!a && !b) return 1;
  if (!a || !b) return 0;
  const d = levenshtein(a, b);
  return 1 - d / Math.max(a.length, b.length, 1);
}

function normalizeLoose(s: string): string {
  return s
    .toLowerCase()
    .replace(/[''`]/g, '')
    .replace(/[^a-z0-9\s]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

function tokenJaccard(a: string, b: string): number {
  const ta = new Set(
    normalizeLoose(a)
      .split(' ')
      .filter((w) => w.length > 2)
  );
  const tb = new Set(
    normalizeLoose(b)
      .split(' ')
      .filter((w) => w.length > 2)
  );
  if (ta.size === 0 && tb.size === 0) return 1;
  if (ta.size === 0 || tb.size === 0) return 0;
  let inter = 0;
  for (const w of ta) {
    if (tb.has(w)) inter++;
  }
  return inter / (ta.size + tb.size - inter);
}

export interface MapRevealOptions {
  /** Minimum composite score to accept a fuzzy match (0–1). */
  minFuzzyScore?: number;
}

/**
 * Returns the canonical `Evidence.title` from `candidates` for one raw model line, or null if no safe match.
 */
export function mapRevealLineToHiddenTitle(
  rawLine: string,
  candidates: EvidenceTitleMatch[],
  opts?: MapRevealOptions
): string | null {
  const minFuzzy = opts?.minFuzzyScore ?? 0.72;
  if (!candidates.length) return null;

  const full = stripRevealedEvidenceMetadata(rawLine).trim();
  if (!full) return null;

  let probe = full;
  if (!/\b(WHERE_HIDDEN|DETAIL)\b/i.test(full) && full.includes(':')) {
    probe = full.split(':')[0].trim();
  }

  const pLow = probe.toLowerCase();
  const fLow = full.toLowerCase();

  for (const e of candidates) {
    const t = e.title.trim();
    if (!t) continue;
    const tl = t.toLowerCase();
    if (tl === pLow || tl === fLow) return t;
  }

  for (const e of candidates) {
    const d = (e.description || '').trim();
    if (d && d.toLowerCase() === fLow) return e.title;
  }

  let bestSub: EvidenceTitleMatch | null = null;
  for (const e of candidates) {
    const tl = e.title.toLowerCase();
    if (tl.length < 4) continue;
    if (pLow.includes(tl) || fLow.includes(tl)) {
      if (!bestSub || e.title.length > bestSub.title.length) bestSub = e;
    }
  }
  if (bestSub) return bestSub.title;

  for (const e of candidates) {
    const tl = e.title.toLowerCase();
    if (tl.length < 4 || pLow.length < 4) continue;
    if (tl.includes(pLow) || pLow.includes(tl)) {
      if (!bestSub || e.title.length > bestSub.title.length) bestSub = e;
    }
  }
  if (bestSub) return bestSub.title;

  let best: EvidenceTitleMatch | null = null;
  let bestScore = 0;
  for (const e of candidates) {
    const t = e.title;
    const lev = Math.max(levSimilarity(pLow, t.toLowerCase()), levSimilarity(fLow, t.toLowerCase()));
    const jac = Math.max(tokenJaccard(probe, t), tokenJaccard(full, t));
    const score = Math.max(lev * 0.55 + jac * 0.45, lev, jac * 0.95);
    if (score > bestScore) {
      bestScore = score;
      best = e;
    }
  }

  if (best && bestScore >= minFuzzy) return best.title;
  return null;
}

/**
 * Map every raw reveal to canonical titles; drop hallucinations; dedupe preserving order.
 */
export function normalizeRevealedEvidenceTitles(
  rawLines: string[],
  unrevealedHidden: EvidenceTitleMatch[],
  opts?: MapRevealOptions
): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const line of rawLines) {
    const canon = mapRevealLineToHiddenTitle(line, unrevealedHidden, opts);
    if (!canon) continue;
    const key = canon.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(canon);
  }
  return out;
}

File: backend/app/services/evidence_reveal_mapping.py
"""
Map model-produced revealedEvidence strings onto canonical hiddenEvidence titles
so chips/collection resolve to real Evidence rows (image + description).
"""
from __future__ import annotations
import re
from dataclasses import dataclass
from typing import Optional


@dataclass
class EvidenceTitleMatch:
    id: str
    title: str
    description: Optional[str] = None


def strip_revealed_evidence_metadata(line: str) -> str:
    s = line.strip()
    idx = re.search(r"\s*\|\s*DISCOVERY_ZONE\b", s, re.IGNORECASE)
    return s[: idx.start()].strip() if idx else s


def _levenshtein(a: str, b: str) -> int:
    if a == b:
        return 0
    m, n = len(a), len(b)
    if m == 0:
        return n
    if n == 0:
        return m
    row = list(range(n + 1))
    for i in range(1, m + 1):
        prev = i - 1
        row[0] = i
        for j in range(1, n + 1):
            tmp = row[j]
            cost = 0 if a[i - 1] == b[j - 1] else 1
            row[j] = min(row[j] + 1, row[j - 1] + 1, prev + cost)
            prev = tmp
    return row[n]


def _lev_similarity(a: str, b: str) -> float:
    if not a and not b:
        return 1.0
    if not a or not b:
        return 0.0
    d = _levenshtein(a, b)
    return 1.0 - d / max(len(a), len(b), 1)


def _normalize_loose(s: str) -> str:
    s = s.lower()
    s = re.sub(r"[''`]", "", s)
    s = re.sub(r"[^a-z0-9\s]", " ", s)
    s = re.sub(r"\s+", " ", s)
    return s.strip()


def _token_jaccard(a: str, b: str) -> float:
    ta = {w for w in _normalize_loose(a).split(" ") if len(w) > 2}
    tb = {w for w in _normalize_loose(b).split(" ") if len(w) > 2}
    if not ta and not tb:
        return 1.0
    if not ta or not tb:
        return 0.0
    inter = len(ta & tb)
    return inter / (len(ta) + len(tb) - inter)


def map_reveal_line_to_hidden_title(
    raw_line: str,
    candidates: list[EvidenceTitleMatch],
    min_fuzzy_score: float = 0.72,
) -> Optional[str]:
    """Returns the canonical Evidence.title from candidates for one raw model line, or None."""
    if not candidates:
        return None

    full = strip_revealed_evidence_metadata(raw_line).strip()
    if not full:
        return None

    probe = full
    if not re.search(r"\b(WHERE_HIDDEN|DETAIL)\b", full, re.IGNORECASE) and ":" in full:
        probe = full.split(":")[0].strip()

    p_low = probe.lower()
    f_low = full.lower()

    # Exact match on title
    for e in candidates:
        t = e.title.strip()
        if not t:
            continue
        tl = t.lower()
        if tl == p_low or tl == f_low:
            return t

    # Exact match on description
    for e in candidates:
        d = (e.description or "").strip()
        if d and d.lower() == f_low:
            return e.title

    # Substring match
    best_sub: Optional[EvidenceTitleMatch] = None
    for e in candidates:
        tl = e.title.lower()
        if len(tl) < 4:
            continue
        if tl in p_low or tl in f_low:
            if not best_sub or len(e.title) > len(best_sub.title):
                best_sub = e
    if best_sub:
        return best_sub.title

    for e in candidates:
        tl = e.title.lower()
        if len(tl) < 4 or len(p_low) < 4:
            continue
        if tl in p_low or p_low in tl:
            if not best_sub or len(e.title) > len(best_sub.title):
                best_sub = e
    if best_sub:
        return best_sub.title

    # Fuzzy match
    best: Optional[EvidenceTitleMatch] = None
    best_score = 0.0
    for e in candidates:
        t = e.title
        lev = max(_lev_similarity(p_low, t.lower()), _lev_similarity(f_low, t.lower()))
        jac = max(_token_jaccard(probe, t), _token_jaccard(full, t))
        score = max(lev * 0.55 + jac * 0.45, lev, jac * 0.95)
        if score > best_score:
            best_score = score
            best = e

    if best and best_score >= min_fuzzy_score:
        return best.title
    return None


def normalize_revealed_evidence_titles(
    raw_lines: list[str],
    unrevealed_hidden: list[EvidenceTitleMatch],
    min_fuzzy_score: float = 0.72,
) -> list[str]:
    """Map every raw reveal to canonical titles; drop hallucinations; dedupe preserving order."""
    seen: set[str] = set()
    out: list[str] = []
    for line in raw_lines:
        canon = map_reveal_line_to_hidden_title(line, unrevealed_hidden, min_fuzzy_score)
        if not canon:
            continue
        key = canon.lower()
        if key in seen:
            continue
        seen.add(key)
        out.append(canon)
    return out
